src/Composer/DependencyResolver/PoolOptimizer.php

```diff
@@ -195,7 +205,7 @@ private function optimizeByIdenticalDependencies(Pool $pool)
                 continue;
             }
 
-            $packageIdsToRemove[$package->id] = true;
+            $this->markPackageForRemoval($package->id);
```

Comment:
Now that you call this function immediately rather than storing the packages in a local temporary array first. Isn't it possible to trigger the irremovable exception which would otherwise have been prevented by a call to keep further down?

Reply:
Not really because of the continue 3 lines above. And irremovablePackages being all added before optimizeByIdenticalDependencies gets called, so I think it's safe. I left the exception in as an extra safeguard but in theory it's not needed.